File: 0019-remove-nth-node-from-end-of-list/two-pointer.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution
{
    public ListNode RemoveNthFromEnd(ListNode head, int n)
    {
        // Create ne wnode to potentially replace the head
        ListNode tmp = new ListNode(0, head);

        // Create two pointers
        ListNode left  = tmp;
        ListNode right = tmp;

        // Move right pointer N steps ahead of left
        for (int i = 0; i < n; i++)
        {
            right = right.next;
        }

        // Move both left and right until right reaches the end
        while (right.next != null)
        {
            left  = left.next;
            right = right.next;
        }

        // Replace the node right after left
        left.next = (left.next).next;

        // It's possible that the original head may be removed
        ListNode res = tmp.next;
        return res;
    }
}

File: 0019-remove-nth-node-from-end-of-list/two-pointer.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        # Create a new node to potentially replace head
        tmp = ListNode(0, head)

        # Create two pointers
        left  = tmp
        right = tmp

        # Move right pointer N steps ahead of left
        for i in range(n):
            right = right.next
        
        # Move both left and right until right reaches the end
        while right.next is not None:
            left  = left.next
            right = right.next
        
        # Replace the node after left
        left.next = (left.next).next

        # It's possible that the original head may be removed
        return tmp.next
        